react-app/src/hooks/useSearchParamsObject.ts

```diff
@@ -0,0 +1,55 @@
+import { useCallback, useEffect, useMemo } from "react";
+import qs from "qs";
+import { useLocation, useSearchParams } from "react-router-dom";
+
+/**
+ * A enhanced useSearchParams that allows objects
+ * rather than only as a key string-value only store
+ *
+ * @param defaultParams - params to set when search string is empty
+ */
+const useSearchParamsObject = <T>(
+  defaultParams?: T
+): [searchParams: T, setSearchParams: (v: T | ((prev: T) => T)) => void] => {
+  const location = useLocation();
+  const [_, setSearchParamsString] = useSearchParams();
+
+  const searchParams: T = useMemo(() => {
+    const searchString = location.search.replace("?", "");
+    const params = qs.parse(searchString, {
+      decoder: (str, defaultDecoder, charset, type) => {
+        if (type === "value") {
+          if (str === "true" || str === "false") {
+            return Boolean(str);
+          }
+          if (!isNaN(parseInt(str, 10))) {
+            return parseInt(str, 10);
+          }
+        }
+        return defaultDecoder(str, null, charset);
+      },
```

Comment:
So the defaultDecoder return everything in string?

Reply:
Yes, see https://github.com/ljharb/qs/blob/95bc0185e157d400da4f43f1fcf1c7f008fd847e/lib/utils.js#L111-L123